### PR TITLE
Made Model.get_value() evaluate states

### PR DESCRIPTION
**Pull request by MichaelClerx**
## Description
See #278 

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)

## Checklist:
- [x] I have updated all documentation necessary.
- [x] I have checked spelling in (new) comments.

## Testing
- [X] Testing is done automatically and codecov shows test coverage

### CODE REVIEW

File: cellmlmanip/model.py [hunk 1]
```diff
@@ -560,8 +560,30 @@ def get_definition(self, variable):
...
+        deps = expr.atoms(VariableDummy)
+        while deps:
+            if evaluated is None:
+                evaluated = {x: x.initial_value for x in self._ode_definition_map.keys()}
+                expr = expr.xreplace(evaluated)
+            for dep in deps:
+                if dep not in evaluated:
+                    evaluated[dep] = self._get_value(dep, evaluated)
+            expr = expr.xreplace(evaluated)
+            deps = expr.atoms(VariableDummy)
```
Comment (jonc125) [inline cellmlmanip/model.py:584]: If `_get_value` is guaranteed to return a float, this should always return an empty set and you never loop. So I think we can get rid of the loop here.

File: cellmlmanip/model.py [hunk 2]
```diff
@@ -560,8 +560,30 @@ def get_definition(self, variable):
...
+
+        return float(expr.evalf())
```
Comment (jonc125) [inline cellmlmanip/model.py:586]:
```suggestion
        return float(expr)
```
Conversion to float calls `evalf` as required.

File: tests/test_model_functions.py [hunk 3]
```diff
@@ -333,9 +333,14 @@ def test_get_definition(self, simple_ode_model):
     def test_get_value(self, aslanidi_model):
         """ Tests Model.get_value() works correctly. """
 
-        symbol_a = aslanidi_model.get_variable_by_ontology_term((shared.OXMETA, "membrane_capacitance"))
+        # Test getting a constant
+        symbol_a = aslanidi_model.get_variable_by_ontology_term((shared.OXMETA, 'membrane_capacitance'))
         assert aslanidi_model.get_value(symbol_a) == 5e-5
 
+        # Test getting an intermediary value
```
Comment (jonc125) [inline tests/test_model_functions.py:340]: Is it worth adding a test that gets a state var directly?

File: cellmlmanip/model.py [hunk 4]
```diff
@@ -560,8 +560,30 @@ def get_definition(self, variable):
         return defn
 
     def get_value(self, variable):
-        """Returns the evaluated value of the given variable's RHS."""
-        return float(self.graph.nodes[variable]['equation'].rhs)
+        """Returns the evaluated value of the given variable, as a float."""
```
Comment (jonc125) [inline cellmlmanip/model.py:563]: I think we should expand this docstring to emphasise that it recursively evaluates referenced equations and uses initial values for state variables.

We could also merge with the internal helper method, and document the extra parameter as for internal use only.

File: tests/test_model_functions.py (hunk 3 above)
Comment (jonc125) [reply tests/test_model_functions.py:340]: Actually, we definitely should, as this case is uncovered!

File: cellmlmanip/model.py (hunk 4 above)
Comment (MichaelClerx) [reply cellmlmanip/model.py:311]: Have updated the docstring, but much happier hiding the argument from the user! It's not part of the public API and also has some hidden constraints: If you were to insert an evaluation for `INa` that would be used, while an evaluation for `V` would not, and if you were to insert evaluations for both states and intermediary values, you'd have to be sure those were consistent or the method behaviour would be undefined

File: cellmlmanip/model.py [hunk 5]
```diff
@@ -308,8 +308,38 @@ def get_definition(self, variable):
...
+
+    def _get_value(self, variable, evaluated=None):
+        """Internal method to implement :meth:`get_value()`."""
+
+        # State? Then return initial value
+        if variable in self._ode_definition_map:
+            return float(variable.initial_value)
+
+        # Get RHS and evaluate
+        expr = self.graph.nodes[variable]['equation']
```
Comment (jonc125) [inline cellmlmanip/model.py:328]: Ooh, actually, we should use
```suggestion
        expr = self._var_definition_map.get(variable)
```
for this now!

File: cellmlmanip/model.py [hunk 6]
```diff
@@ -308,8 +308,38 @@ def get_definition(self, variable):
...
+        if expr is None:
+            return 0
```
Comment (jonc125) [inline cellmlmanip/model.py:330]: Wouldn't it be better to raise a sensible error for this? More like an undefined variable case isn't it?

Comment (jonc125) [reply cellmlmanip/model.py:330]: Ah, I see from the test below this is to support time. Probably better to handle time explicitly; see suggestion below?

File: cellmlmanip/model.py [hunk 7]
```diff
@@ -308,8 +308,38 @@ def get_definition(self, variable):
...
+        expr = expr.rhs
+        deps = expr.atoms(VariableDummy)
+        if deps:
+            if evaluated is None:
+                evaluated = {x: x.initial_value for x in self._ode_definition_map.keys()}
```
Comment (jonc125) [inline cellmlmanip/model.py:335]:
```suggestion
                evaluated = {x: x.initial_value for x in self._ode_definition_map.keys()}
                if self._ode_definition_map:
                    time = self.get_free_variable()
                    evaluated[time] = 0
```

**Comment by MichaelClerx** [comment]
Sorry did another slight update @jonc125 :D

**Pull request merged**